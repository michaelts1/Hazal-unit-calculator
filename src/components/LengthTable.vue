<script lang="ts">
import { defineComponent, PropType } from 'vue'
import { roundNum } from '../helpers'
import type { Unit } from '../types'

export default defineComponent({
	props: {
		units: {
			type: Array as PropType<Unit[]>,
			required: true,
		},
	},

	methods: {
		/**
		 * Formats unit measurements
		 * @param {number} num
		 * @returns {string}
		 */
		formatNum(num: number): string {
			let unit = ' מטר'

			if (num < .01) {
				num *= 1000
				unit = ' מ"מ'
			} else if (num < 1) {
				num *= 100
				unit = ' ס"מ'
			} else if (num > 1000) {
				num /= 1000
				unit = ' ק"מ'
			}

			return roundNum(num) + unit
		},
	},
})
</script>

<template>
	<table>
		<thead>
			<tr>
				<th
					v-for="{ name } in units"
					:key="name"
				>
					{{ name }}
				</th>
			</tr>
		</thead>
		<tbody>
			<tr>
				<td
					v-for="{ value } in units"
					:key="value"
				>
					{{ formatNum(value) }}
				</td>
			</tr>
		</tbody>
	</table>
</template>

<style scoped>
	.centered {
		margin: 0 auto;
		text-align: center;
	}

	.wrapper-row {
		display: flex;
		margin: 1em auto;
		width: fit-content;
	}

	.wrapper-column {
		align-items: center;
		display: flex;
		flex-direction: column;
		justify-content: center;
	}
</style>
