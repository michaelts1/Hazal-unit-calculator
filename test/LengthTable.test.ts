import { describe, expect, it } from 'vitest'
import LengthTable from '../src/components/LengthTable.vue'
import { mount } from '@vue/test-utils'
import type { Unit } from '../src/types'

interface ColumnData {
	unit: string
	value: number
}

describe('Component LengthTable', () => {
	it('exists', () => {
		expect(LengthTable).toBeTruthy()
	})

	const units: Unit[] = [
		{ hidden: false, name: 'אצבע', value: 0.005 },
		{ hidden: false, name: 'סנטימטר', value: 0.1 },
		{ hidden: true, name: 'אמה', value: 2 },
		{ hidden: false, name: 'טפח', value: 4 },
		{ hidden: false, name: 'זרת', value: 4000 },
	]
	const wrapper = mount(LengthTable, {
		props: {
			units,
		},
	})

	it('has a table with the correct number of columns', () => {
		expect(wrapper.find('table').exists()).toBe(true)
		expect(wrapper.findAll('th')).toHaveLength(units.filter(u => !u.hidden).length)
	})

	it('matches snapshot', () => {
		expect(wrapper.html()).toMatchSnapshot()
	})

	it('lists the correct values and reacts to value changes', async () => {
		expect(wrapper.text()).toContain('5 מ"מ')
		expect(wrapper.text()).not.toContain('2 מטר')
		expect(wrapper.text()).toContain('4 ק"מ')

		units[0].value = 0.5
		await wrapper.vm.$forceUpdate() // $forceUpdate is async (the type definition is wrong)
		expect(wrapper.text()).not.toContain('5 מ"מ')
		expect(wrapper.text()).toContain('50 ס"מ')
	})

	it('responds to events emitted by SelectUnit', () => {
		const [ input, output ] = wrapper.findAllComponents({ name: 'SelectUnit' })

		expect((wrapper.vm.input as ColumnData).unit).toBe('טפח') // Default
		input.vm.$emit('value-change', 'אצבע')
		expect((wrapper.vm.input as ColumnData).unit).toBe('אצבע') // New value

		expect((wrapper.vm.output as ColumnData).unit).toBe('סנטימטר') // Default
		output.vm.$emit('value-change', 'אמה')
		expect((wrapper.vm.output as ColumnData).unit).toBe('אמה') // New value
	})

	it('responds to events emitted by InputAmount', () => {
		const input = wrapper.findComponent({ name: 'InputAmount' })

		expect((wrapper.vm.input as ColumnData).value).toBe(0) // Default
		input.vm.$emit('value-change', 10)
		expect((wrapper.vm.input as ColumnData).value).toBe(10) // New value
	})
})
