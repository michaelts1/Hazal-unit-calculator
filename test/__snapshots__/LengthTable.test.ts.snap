// Vitest Snapshot v1

exports[`Component LengthTable > matches snapshot 1`] = `
"<table data-v-4279aa88=\\"\\">
  <thead data-v-4279aa88=\\"\\">
    <tr data-v-4279aa88=\\"\\">
      <th data-v-4279aa88=\\"\\">אצבע</th>
      <th data-v-4279aa88=\\"\\">סנטימטר</th>
      <th data-v-4279aa88=\\"\\">טפח</th>
      <th data-v-4279aa88=\\"\\">זרת</th>
    </tr>
  </thead>
  <tbody data-v-4279aa88=\\"\\">
    <tr data-v-4279aa88=\\"\\">
      <td data-v-4279aa88=\\"\\">5 מ\\"מ</td>
      <td data-v-4279aa88=\\"\\">10 ס\\"מ</td>
      <td data-v-4279aa88=\\"\\">4 מטר</td>
      <td data-v-4279aa88=\\"\\">4 ק\\"מ</td>
    </tr>
  </tbody>
</table>
<div class=\\"wrapper-row\\" data-v-4279aa88=\\"\\">
  <div class=\\"wrapper-column\\" data-v-4279aa88=\\"\\"><select class=\\"centered\\" data-v-4279aa88=\\"\\">
      <option value=\\"אצבע\\">אצבע</option>
      <option value=\\"סנטימטר\\">סנטימטר</option>
      <option value=\\"אמה\\">אמה</option>
      <option value=\\"טפח\\">טפח</option>
      <option value=\\"זרת\\">זרת</option>
    </select><input class=\\"conversion-input\\" size=\\"12\\" data-v-519321c2=\\"\\" data-v-4279aa88=\\"\\"></div>
  <div class=\\"wrapper-column\\" data-v-4279aa88=\\"\\"> = </div>
  <div class=\\"wrapper-column\\" data-v-4279aa88=\\"\\"><select class=\\"centered\\" data-v-4279aa88=\\"\\">
      <option value=\\"אצבע\\">אצבע</option>
      <option value=\\"סנטימטר\\">סנטימטר</option>
      <option value=\\"אמה\\">אמה</option>
      <option value=\\"טפח\\">טפח</option>
      <option value=\\"זרת\\">זרת</option>
    </select><input class=\\"conversion-input\\" disabled=\\"\\" size=\\"12\\" data-v-519321c2=\\"\\" data-v-4279aa88=\\"\\"></div>
</div>"
`;
